Added Howell form and strong echelon form for nmod_mat

thofma opened the pull request:
We had the code for the last few months in fieker/antic and use it in hecke for Hermite normal form computations. I have now added tests and documentation.

The Howell form is derived from the so-called strong echelon form, see [Fieker, Hofmann, Computing in quotients of ring of integers](http://dx.doi.org/10.1112/S1461157014000291). The algorithm is essentially the classical algorithm of Storjohann & Mulders.

The functions have the restriction that the matrix should have more rows then columns. This is no real restriction as one can just add zero-rows at the bottom. Since this has a bad memory footprint, I don't do it in the algorithm. Probably there is clever way I didn't found.

I have a correspdoning fmpz_mat_mod_howell_form ready (apart from test and documentation). Once I have merged both I will add a modular Hermite normal form which accepts a multiple of the largest elementary divisor (this is the overall goal).

I know that the implementation can be improved, but I don't have time for it (at the moment).

Edit: The failing travis test is due to mpfr.org begin down (not my fault!).

wbhart commented:
This all looks fine to me. However, can I ask for one very small change before I merge it. Can you add a reference to a paper describing the Howell form to the documentation.

Also, what is strong echelon form? A description of what it is mathematically would help in the documentation for people like me who haven't heard of it.

thofma commented:
I added some references to the documentation and fixed a bug in the signature.

The strong echelon form is equal to the Howell form up to permutation of the rows, e.g., if ```A``` has Howell form
```
0 2 1
0 0 2
0 0 0
```
then the strong echelon form of ```A``` is
```
0 0 0
0 2 1
0 0 2
```
While this looks trivial, this allows for a modular computation of the strong echelon form in the following sense: If ```a,b``` are coprime integers and ```A``` is a matrix over ```Z/abZ```, then the strong echelon form of ```A``` can be computed form the strong echelon form of ```A mod a``` and ```A mod b``` using CRT.

Second advantage: If ```A``` is an integer matrix and ```a``` a multiple of the largest elementary divisor, then the Hermite form of ```A``` can be read off the strong echelon form of ```A mod a```.

Consequences for the Hermite normal form computation of an integer matrix:
- You can use a multiple ```a``` of the elementary divisor instead of the determinant.
- If you know a factorization ```a = bc``` into coprime integers, you can use the modular technique.

For random integer matrices this won't help you (largest elementary divisor and determinant differ only by 2 and cannot by factored). Nevertheless for the matrices in our applications (ideals and orders in number fields), the largest elementary divisor and some factorization of it is readily available.

Hope this helps!
